Migrate Ansible tower modules to collection maintained alongside AWX

AlanCoding opened the pull request:
##### SUMMARY
PR to replace https://github.com/ansible/awx/pull/4556

Right now this is to confirm tests are running.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
7.0.0
```


##### ADDITIONAL INFORMATION
Still TODO:

 - fix doc fragments (they reference the version in Ansible core)
 - TBD: rename `tower_organization` to just `organization`
 - squash all of my own commits
 - probably more stuff I am not thinking of right now

after merge TODO:

 - handle open PRs in Ansible core
 - upload
 - create label for associated issues
 - deprecate duplicated content in Ansible core

----
Cherry pick list

https://github.com/ansible/ansible/pull/62202
https://github.com/ansible/ansible/pull/61338
https://github.com/ansible/ansible/pull/62157

Commit:
Ansible Tower organization module (#20355)

* add Ansible Tower organization module

* skip Python 2.4 check for ansible_tower module

* make spec and doc match, extract tower auth helper method

* added auth params at module level

* support check mode

* extract check mode check to ansible_tower utils, add utils to 2.4 skip

* update interpreter shebang

* remove colon from docs

* no log for password, verify_ssl default to true

bcoca committed:
updated friendlier description

sivel committed:
Update validate-modules (#20932)

* Update validate-modules

* Validates ANSIBLE_METADATA
* Ensures imports happen after documentation vars
* Some pep8 cleanup

* Clean up some left over unneeded code

* Update modules for new module guidelines and validate-modules checks

* Update imports for ec2_vpc_route_table and ec2_vpc_nat_gateway

Commit:
Ansible Tower user and credential module (#21020)

* rename tower config module parameters to avoid conflicts

* add Ansible Tower user module

* add Ansible Tower credential module

* remove errant hash from interpreter line

* friendlier error messages

* Update tower_verify_ssl defaults and module examples

* Update tower_verify_ssl default documentation

* Tower expects satellite6 not foreman

Commit:
Add Tower Label module (#21485)

Commit:
Add Tower Inventory module (#21483)

Commit:
Add Tower Host module (#21482)

Commit:
Add Tower Group module (#21480)

Commit:
Add Tower Project module (#21479)

gundalow committed:
Correct example

Commit:
add Ansible Tower team module (#21593)

Commit:
add Tower JobTemplate module (#21681)

* add Tower JobTemplate module

* add host_config_key and remove defaults from required parameters

Commit:
add Ansible Tower role module (#21592)

* add Ansible Tower role module

* remove owner as choice from role paramenter

Commit:
Ansible Tower job_launch module (#22148)

* Ansible Tower job_launch module

* Added RETURN documentation and fixed import locations

* remove superfluos required attributes, make tags a list, and fix some typos

* only join tags if they are actually a list

* use isinstance instead of type, cleanup imports

Commit:
Ansible Tower job list module (#22164)

Commit:
Ansible Tower job cancel module (#22161)

* Ansible Tower job cancel module

* fix interpreter line

Commit:
Ansible Tower job_wait module (#22160)

* Ansible Tower job_wait module

* clean up documentation and update code comment

gundalow committed:
Fix invalid fields in module DOCUMENATION (#22297)

* fix module doc fields

* More module docs corrections

* More module docs corrections

* More module docs corrections

* More module docs corrections

* correct aliases

* Review comments

* Must quote ':'

* More authors

* Use suboptions:

* restore type: bool

* type should be in the same place

* More tidyups

* authors

* Use suboptions

* revert

* remove duplicate author

* More issues post rebase

abadger committed:
New metadata 1.0 (#22587)

Changes to the metadata format were approved here:
https://github.com/ansible/proposals/issues/54
* Update documentation to the new metadata format
* Changes to metadata-tool to account for new metadata
  * Add GPL license header
  * Add upgrade subcommand to upgrade metadata version
  * Change default metadata to the new format
  * Fix exclusion of non-modules from the metadata report
* Fix ansible-doc for new module metadata
* Exclude metadata version from ansible-doc output
* Fix website docs generation for the new metadata
* Update metadata schema in valiate-modules test
* Update the metadata in all modules to the new version

Shaps committed:
fixed RETURN docs for modules (#24011)

* fixed RETURN docs for remaining modules

* updated proxymysql_mysql_users 'sample' to yaml dict

* fixed whitespace errors

Akasurde committed:
Pep8 fixes for web_infra/ansible_tower (#24479)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

galtenberg committed:
Improve help text for extra-vars requiring @ for filename

(cherry picked from commit 1b34de89ee1d75cb7f616b5a34cd5043bf7dfd2b)

lshake committed:
Tower user remove organization (#24544)

* Remove organization field from the tower_user module re: issue #24510

* Fix trailing spaces.

* Fixes for Shippable errors, pep8

* Remove a random inserted space.

lshake committed:
Lookup credential id and pass in credential rather than scm_credential (#24624)

* Lookup credential id and pass in credential rather than scm_credential

* Change the excepting handling to catch missing credentials

* Make error messages for not found lookups more useful

kevclarx committed:
adds privilege escalation method for pmrun(Unix Privilege Manager 6.0)

jameslabocki committed:
Fix indentation for register module in example (#25274)

dagwieers committed:
Collated PEP8 fixes (#25293)

- Make PEP8 compliant

wenottingham committed:
Fix handling of extra_vars_path parameter. (#25272)

tower-cli process_extra_vars takes a list.

Akasurde committed:
Fix spelling mistakes (comments only) (#25564)

Original Author : klemens <ka7@github.com>

Taking over previous PR as per
https://github.com/ansible/ansible/pull/23644#issuecomment-307334525

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

abadger committed:
Remove wildcard imports

Made the following changes:

* Removed wildcard imports
* Replaced long form of GPL header with short form
* Removed get_exception usage
* Added from __future__ boilerplate
  * Adjust division operator to // where necessary

For the following files:

* web_infrastructure modules
* system modules
* linode, lxc, lxd, atomic, cloudscale, dimensiondata, ovh, packet,
  profitbricks, pubnub, smartos, softlayer, univention modules
* compat dirs (disabled as its used intentionally)

abadger committed:
Update metadata to 1.1

ethackal committed:
Fixes verify_ssl option when False in ansible_tower module util (#30308)

* Fixes verify_ssl option when False in ansible_tower module util

* fixed comparison to None per PEP-8 standards

pilou- committed:
ansible_tower: fix broken import, reuse tower_argument_spec and documentation fragment (#29115)

* module_utils/ansible_tower: fix broken import

* tower_*: use tower_argument_spec & doc fragment

* tower doc fragment: Ansible requires Python 2.6+

* tower_job_wait: fix broken import (Py3 compat)

abadger committed:
Port arg specs from type='str' to type='path'

tbouvet committed:
Fix credentials for Tower API V2

ryanpetrello committed:
tower cred: implement credential /api/v1/ kind compatability

ryanpetrello committed:
tower cred: filter user name lookup by the proper key

ryanpetrello committed:
tower cred: update kind options in documentation

ryanpetrello committed:
tower cred: support credential kind/type for /api/v1/ and /api/v2/ (#36662)

older versions of Tower (3.1) don't have a concept of CredentialTypes
(this was introduced in Tower 3.2).  This change detects older versions
of pre-3.2 tower-cli that *only* support the deprecated `kind`
attribute.

ryanpetrello committed:
properly detect the absence of credential_type in older tower-cli (#36908)

ryanpetrello committed:
properly pass /api/v1/ credential fields for older Towers (#36917)

pilou- committed:
ansible_tower modules doc: fix typos, use formatting functions (#37414)

* fix typos

* use formatting functions

* use 'job template' instead of 'job_template'

* acronyms: user uppercase

* become_enabled param is about privilege escalation

dagwieers committed:
Clean up module documentation (#36909)

* Clean up module documentation

This PR includes:
- Removal of `default: None` (and variations)
- Removal of `required: false`
- Fixing booleans and `type: bool` where required

* Fix remaining (new) validation issues

ryanpetrello committed:
add some Tower module integration tests (and fix a bug or two) (#37421)

* add additional test coverage for tower modules

* add test coverage for the tower_credential module

* add test coverage for the tower_user module

* fix a bug in py3 for tower_credential when ssh_key_data is specified

* add test coverage for tower_host, tower_label, and tower_project

* add test coverage for tower_inventory and tower_job_template

* add more test coverage for tower modules

- tower_job_launch
- tower_job_list
- tower_job_wait
- tower_job_cancel

* add a check mode/version assertion for tower module integration tests

* add test coverage for the tower_role module

* add test coverage for the tower_group module

* add more integration test edge cases for various tower modules

* give the job_wait module more time before failing

* randomize passwords in the tower_user and tower_group tests

piroux committed:
Fix tower_* modules **params kwargs (#40137)

* Add cleaning function to handle **params

The cleaning function is only added to tower modules which pass a `**params`
argument as an unpacked dictionnary to the tower-cli method calls.

Fix #39745

* Remove previous code added only for tower_role

In 872a7b4, the `update_resources` function was modified so that it would clear unwanted
parameters. However, this behaviour is desired for other modules too, modified in
another commit. (see tower_clean_params).

fleu42 committed:
Module: Tower inventory source module (#37110)

* tower_inventory_source: Add support for the inventory source via ansible-tower-cli.

* Add test coverage for tower_inventory_source.

* Update version_added to 2.7

Commit:
Updating tower_job_template.py (#38821)

* Updating tower_job_template.py

* tower_job_template: Update parameter version_added to 2.7

* Ensure that unset credentials aren't passed

Passing empty strings for unset credentials causes ValueErrors as
the API expects an integer. Don't pass unset credentials

jaevans committed:
Add support for Tower Smart inventories (#41458)

* Support Smart Inventories

Add kind and host_filter fields and pass through to tower_cli.

* Add documentation for new Smart Inventories options

* Add missing description header for host_filter documentation

* Add version added tags to new options

* Bumped vesion_added to 2.7

Akasurde committed:
Correct typo from 'Valut' to 'Vault' (#41574)

Correct typo from 'Valut' to 'Vault'

+label: docsite_pr

pilou- committed:
tower_project: manual projects don't require creds

Commit:
Properly detect credentials for tower_project

It seemed like it was mostly the wrong variables been looked at, making
it so a git repository could not be created without a credential.

pilou- committed:
Tower modules: move HAS_TOWER_CLI check in module_utils and minor improvements (#39809)

* tower_* modules: move HAS_TOWER_CLI in TowerModule

Besides this change allows to define other common parameters such as
mutually_exclusive.

* tower_*: config file can not be used with auth params

* tower module_utils: remove useless call to expanduser

'path' type: expanduser & expandvars are automatically called

willthames committed:
Allow tower_inventory_sources params to be False (#43749)

Ensure that params set to False are respected, rather than ignored.

curry9999 committed:
To improve readability, we added a line feed. (#43801)

* 	modified:   google/gcp_compute_backend_bucket.py
	modified:   google/gcp_compute_backend_service.py
	modified:   google/gcp_compute_forwarding_rule.py
	modified:   google/gcp_compute_global_forwarding_rule.py
	modified:   google/gcp_compute_image.py
	modified:   google/gcp_compute_instance.py
	modified:   google/gcp_compute_instance_group.py
	modified:   google/gcp_compute_instance_group_manager.py
	modified:   google/gcp_compute_instance_template.py
	modified:   google/gcp_compute_route.py
	modified:   google/gcp_compute_subnetwork.py
	modified:   google/gcp_compute_target_http_proxy.py
	modified:   google/gcp_compute_target_https_proxy.py
	modified:   google/gcp_compute_target_ssl_proxy.py
	modified:   google/gcp_compute_target_tcp_proxy.py
	modified:   google/gcp_compute_url_map.py
	modified:   google/gcp_container_node_pool.py
	modified:   google/gcp_dns_resource_record_set.py
	modified:   google/gcp_pubsub_subscription.py
	modified:   google/gcp_storage_bucket_access_control.py

* 	modified:   lib/ansible/modules/cloud/amazon/aws_ses_identity.py
	modified:   lib/ansible/modules/cloud/amazon/route53_facts.py
	modified:   lib/ansible/modules/cloud/cloudscale/cloudscale_server.py
	modified:   lib/ansible/modules/network/aos/_aos_logical_device.py
	modified:   lib/ansible/modules/network/aos/_aos_rack_type.py
	modified:   lib/ansible/modules/network/aos/_aos_template.py
	modified:   lib/ansible/modules/network/cumulus/nclu.py
	modified:   lib/ansible/modules/web_infrastructure/ansible_tower/tower_job_launch.py
	modified:   lib/ansible/modules/web_infrastructure/ansible_tower/tower_job_wait.py

YunfanZhang42 committed:
Added Ansible Tower inventory plugin. (#41816)

Signed-off-by: Yunfan Zhang <yz322@duke.edu>

jainnikhil30 committed:
Add 'tower_settings' module for managing Ansible Tower Settings (#43933)

* add the tower_settings module

fleu42 committed:
New module tower_workflow_template. (#37520)

* Add new module *tower_workflow_template*

Manage Tower workflows and their schemas.

fleu42 committed:
New module : tower_credential_type (#37243)

* Add new module *tower_credential_type*

* Add support for credential type creation and/or deletion
* Add test coverage for tower_credential_type

Meecr0b committed:
tower_credential: expect ssh_key_data to be a string instead of path (#45158)

* expect ssh_key_data to be a string instead of path

ssh_key_data should be a string filled with the private key
the old behavior can be archived with a lookup

Fixes #45119

* clarifies ssh_key_data description, adds newline

dagwieers committed:
Docs: Avoid use of 'default: null' (#45795)

Various modules document the default 'null' value, but it causes None to
be shown in the documentation explicitly.

Spredzy committed:
inventory/tower: authors -< author so Doc can pick it up (#45936)

s-hertel committed:
add more consistent extension matching for inventory plugins (#46786)

* Add consistent extension matching for inventory plugins that support YAML configuration files

* Document extension matching expectations

IPvSean committed:
adding additional example and clarification (#47224)

The kind: paramter is missing from the original example (which won't actually work).  The organization needs to be listed as required (if you don't list it, you will get an error).  Also adding additional example using the tower_username and tower_password method.

bbyhuy committed:
Add survey_spec parameter to module. (#48182)

* Add survey_spec parameter to module.
Fixes #48011

* Removed trailing white space. Added integration test.

seag-rvc committed:
Update tower_host.py (#49186)

<!--- Your description here -->
Current example does not show how to declare variables
+label: docsite_pr

Akasurde committed:
E325 Removal - Part II (#49196)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

dagwieers committed:
Convert to reduced list of known types (#50010)

Commit:
Added organization in the scm_credential get (#49884)

* Added organization in the scm_credential get

* Fallback looking for cred in project org

* Tests project with multi org credential

* Fixed CI issue

* Added changelog fragment

acozine committed:
fix docs for tower modules (#50710)

jainnikhil30 committed:
Fixing exception import for tower modules (#50447)

* fixing the exception import from tower modules

* Adding tests for checking tower modules are failing with correct msg

* fixed failing tests

* fixed failing test in tower_team

john-westcott-iv committed:
Adding TowerCLI send module (#37843)

john-westcott-iv committed:
Adding tower_workflow_launch (#42701)

john-westcott-iv committed:
Adding TowerCLI receive module (#51023)

jainnikhil30 committed:
Add scm_update_cache_timeout, job_timeout and custom_virtualenv to tower_project (#51330)

* adding scm_update_cache_timeout and job_timeout to tower_project module

* add support for cache_timeout, job_timeout and custom_venv for tower_project module

* add the version_added

samcarpentier committed:
New module: tower_notification (#50512)

* New module: tower_notification

* Fix CI check failures

* Add integration tests and extend examples

* Add missing required field for deletion tests and examples

* Add missing required field for deletion tests and examples

* Set port type to int

* Add missing field for Slack notification

* Add missing field types for IRC notification

* Update module documentation

* Correct field name and type for IRC notification

* Uniformize 'targets' field

* Uniformize 'targets' field

dlerch-transporeon committed:
add diff_mode_enabled to documentation

option 'diff_mode_enabled' is not mentioned in documentation

+label: docsite_pr

jborean93 committed:
Final round of moving modules to new import error msg (#51852)

* Final round of moving modules to new import error msg

* readd URL to jenkins install guide

* fix unit tests

pilou- committed:
tower modules: check that 'verify_ssl' defined in ~/.tower_cli.cfg isn't ignored (#50687)

* Check that verify_ssl defined in tower_cli.cfg isn't ignored

* Avoid to override verify_ssl value defined in tower_cli.cfg

By default, tower-cli library enables SSL certificates check. But
verify_ssl false value defined in config files read by default by
tower-cli library (for example /etc/tower/tower_cli.cfg) was ignored
because overriden by the tower_verify_ssl parameter default value.

* fix a typo in comment

medberry committed:
Update tower_credential.py (#51469)

misspelled word and poor capitalization (in the same sentence of the doc string)

+label: docsite_pr

Akasurde committed:
tower: Handle AuthError (#53377)

Handle AuthError raised when user provides incorrect password
for Tower admin user.

Fixes: #50535

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

AlanCoding committed:
Add option for tower inventory to give general metadata (#52747)

Akasurde committed:
tower_credential: Add parameter vault_id (#53400)

vault_id allows user to specify vault identifier as per Tower UI.

Fixes: #45644

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

pilou- committed:
tower_settings: "get" isn't implemented, "value" parameter is required (#54028)

* tower_settings doc: 'get' isn't implemented

* tower_settings: fix typo in argument_spec

jamescassell committed:
standardize TLS connection properties (#54315)

* openstack: standardize tls params

* tower: tower_verify_ssl->validate_certs

* docker: use standard tls config params

- cacert_path -> ca_cert
- cert_path -> client_cert
- key_path -> client_key
- tls_verify -> validate_certs

* k8s: standardize tls connection params

- verify_ssl -> validate_certs
- ssl_ca_cert -> ca_cert
- cert_file -> client_cert
- key_file -> client_key

* ingate: verify_ssl -> validate_certs

* manageiq: standardize tls params

- verify_ssl -> validate_certs
- ca_bundle_path -> ca_cert

* mysql: standardize tls params

- ssl_ca -> ca_cert
- ssl_cert -> client_cert
- ssl_key -> client_key

* nios: ssl_verify -> validate_certs

* postgresql: ssl_rootcert -> ca_cert

* rabbitmq: standardize tls params

- cacert -> ca_cert
- cert -> client_cert
- key -> client_key

* rackspace: verify_ssl -> validate_certs

* vca: verify_certs -> validate_certs

* kubevirt_cdi_upload: upload_host_verify_ssl -> upload_host_validate_certs

* lxd: standardize tls params

- key_file -> client_key
- cert_file -> client_cert

* get_certificate: ca_certs -> ca_cert

* get_certificate.py: clarify one or more certs in a file

Co-Authored-By: jamescassell <code@james.cassell.me>

* zabbix: tls_issuer -> ca_cert

* bigip_device_auth_ldap: standardize tls params

- ssl_check_peer -> validate_certs
- ssl_client_cert -> client_cert
- ssl_client_key -> client_key
- ssl_ca_cert -> ca_cert

* vdirect: vdirect_validate_certs -> validate_certs

* mqtt: standardize tls params

- ca_certs -> ca_cert
- certfile -> client_cert
- keyfile -> client_key

* pulp_repo: standardize tls params

remove `importer_ssl` prefix

* rhn_register: sslcacert -> ca_cert

* yum_repository: standardize tls params

The fix for yum_repository is not straightforward since this module is
only a thin wrapper for the underlying commands and config.  In this
case, we add the new values as aliases, keeping the old as primary,
only due to the internal structure of the module.

Aliases added:
- sslcacert -> ca_cert
- sslclientcert -> client_cert
- sslclientkey -> client_key
- sslverify -> validate_certs

* gitlab_hook: enable_ssl_verification -> hook_validate_certs

* Adjust arguments for docker_swarm inventory plugin.

* foreman callback: standardize tls params

- ssl_cert -> client_cert
- ssl_key -> client_key

* grafana_annotations: validate_grafana_certs -> validate_certs

* nrdp callback: validate_nrdp_certs -> validate_certs

* kubectl connection: standardize tls params

- kubectl_cert_file -> client_cert
- kubectl_key_file -> client_key
- kubectl_ssl_ca_cert -> ca_cert
- kubectl_verify_ssl -> validate_certs

* oc connection: standardize tls params

- oc_cert_file -> client_cert
- oc_key_file -> client_key
- oc_ssl_ca_cert -> ca_cert
- oc_verify_ssl -> validate_certs

* psrp connection: cert_trust_path -> ca_cert

TODO: cert_validation -> validate_certs (multi-valued vs bool)

* k8s inventory: standardize tls params

- cert_file -> client_cert
- key_file -> client_key
- ca_cert -> ca_cert
- verify_ssl -> validate_certs

* openshift inventory: standardize tls params

- cert_file -> client_cert
- key_file -> client_key
- ca_cert -> ca_cert
- verify_ssl -> validate_certs

* tower inventory: verify_ssl -> validate_certs

* hashi_vault lookup: cacert -> ca_cert

* k8s lookup: standardize tls params

- cert_file -> client_cert
- key_file -> client_key
- ca_cert -> ca_cert
- verify_ssl -> validate_certs

* laps_passord lookup: cacert_file -> ca_cert

* changelog for TLS parameter standardization

saito-hideki committed:
Fix handling of inventory and credential options for tower_job_launch (#54967)

- Fixed issue #25017,#37567
- Add example for prompt on launch
- Add integration test for prompt on launch

Signed-off-by: Hideki Saito <saito@fgrep.org>

saito-hideki committed:
Fixed wrong variable specification format in examples (#55252)

Signed-off-by: Hideki Saito <saito@fgrep.org>

stoned committed:
Add missing roles to tower_role module (#56182)

* Add missing roles to tower_role module

* Placate 'ansible-test sanity --test pep8'

pilou- committed:
tower_credential: ssh_key_data isn't a path anymore (#57113)

saito-hideki committed:
tower_user: Fix to create a system auditor properly (#54585)

- Fixed issue #54446: tower_user cannot create an auditor user

Signed-off-by: Hideki Saito <saito@fgrep.org>

pilou- committed:
fix tower_credential example: lookup are run on controller (#57516)

- don't advise to use lookups in order to fetch path on managed nodes
- fix example:
    - use slurp instead 'file' lookup
    - remove extraneous brackets

albertomurillo committed:
tower_workflow_template: Add missing options (#56891)

The following options can be set on a workflow template but the
functionallity to do so from this module was missing.

inventory
ask_variables_on_launch
ask_inventory_on_launch

Fixes #49728

Signed-off-by: Alberto Murillo <albertomurillosilva@gmail.com>

pilou- committed:
tower_role: ensure alias of "validate_certs" parameter is handled (#57518)

* tower_role: ensure alias of validate_certs is handled

* tower modules: remove tower_verify_ssl alias too

Error was:

    Failed to update role: The Tower server claims it was sent a bad request.
    GET https://tower/api/v2/projects/22/object_roles/
    Params: [('tower_verify_ssl', False), ('role_field', 'admin_role')]
    Data: None
    Response: {"detail": "Role has no field named 'tower_verify_ssl'"}

Full traceback:

    File "/tmp/ansible_tower_role_payload_7_2p0X/__main__.py", line 145, in main
      result = role.grant(**params)
    File "/usr/local/lib/python2.7/dist-packages/tower_cli/resources/role.py", line 365, in grant
      return self.role_write(fail_on_found=fail_on_found, **kwargs)
    File "/usr/local/lib/python2.7/dist-packages/tower_cli/resources/role.py", line 242, in role_write
      fail_on_multiple_results=True, **data)
    File "/usr/local/lib/python2.7/dist-packages/tower_cli/models/base.py", line 301, in read
      r = client.get(url, params=params)
    File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 546, in get
      return self.request('GET', url, **kwargs)
    File "/usr/local/lib/python2.7/dist-packages/tower_cli/api.py", line 299, in request
      kwargs.get('data', None), r.content.decode('utf8'))

acozine committed:
clarifies how ASK works for Tower credentials (#59050)

* clarifies how ASK works for credentials

AlanCoding committed:
Move commit for migration of Ansible core tower modules

AlanCoding committed:
Integrate Ansible core tower modules content into AWX

This commit includes all the changes involved in
converting the old Ansible Tower modules from commits
in Ansible core into the AWX collection that replaces it.
Also includes work needed to integrate it into the
AWX processes like tests, docs, and the Makefile.

Apply changes from content_collector tool

Add integrated module tests
  operate via run_module fixture
  add makefile target for them

Add flake8 target and fix flake8 errors

Update README

Make consolidated target for testing modules

AlanCoding commented:
Put up dry run at:

https://galaxy.ansible.com/alancoding/awx

AlanCoding commented:
recheck

ryanpetrello commented on awx_modules/test/awx/conftest.py:
```diff
@@ -0,0 +1,113 @@
+
+
+@pytest.fixture
+def run_module():
```
This is simultaneously impressive and horrifying.

AlanCoding commented:
I wanted to save some timing information from running these locally. I put in `@pytest.mark.parametrize('foo', range(100))` to a test so it'll run 100 times.

```
== 101 passed, 1 xfailed in 107.22 seconds ==

real	2m21.936s
user	1m18.230s
sys	0m6.590s
```

Compare to `time py.test awx/main/tests/functional/`.

```
== 1466 passed, 7 skipped in 950.56 seconds ==

real	16m11.597s
user	9m41.780s
sys	0m43.160s
```

That comes out to about 1.4 vs 0.6 seconds per test. That gap narrows a lot if you use the reported time as opposed to "real" time, like 1.0 vs. 0.6.

AlanCoding committed:
Allow tower inventory plugin to accept integer inventory_id (#61338)

AlanCoding committed:
Use to_native for error messages, fix docs typo

ryanpetrello commented on awx_modules/plugins/inventory/tower.py:
```diff
@@ -0,0 +1,204 @@
+# Copyright (c) 2018 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
@wenottingham just want to confirm this is kosher since AWX itself isn't importing and running this code (we're just including it in the source tree)

wenottingham replied:
the scripts we vendor likely already have this

AlanCoding replied:
I consulted with @ghjm about this too. I'm planning to push a commit that moves:

https://github.com/ansible/ansible/blob/devel/COPYING

to `awx_modules/COPYING` or `awx_modules/LICENSE`.

I think that's the right thing to do. I highly suspect we also need some sort of notice in the main AWX README indicating that not all folders are under the same license.

@ryanpetrello I would assume that the folder `awxkit` has the same licensing as the rest of AWX.

Andersson007 committed:
fix typos in web_infrastructure modules (#62202)

elyezer commented on Makefile:
```diff
@@ -377,8 +378,26 @@ test:
+	fi; \
+	PYTHONPATH=$(MODULES_VENV):/awx_devel/awx_modules:$PYTHONPATH py.test $(MODULES_TEST_DIRS)
+
+prepare_test_modules: prepare_modules_venv test_modules
```
This is confusing to me, the target is named prepare but it is actually running the tests. I would suggest moving the `prepare_modules_venv` as a dependency of test_modules and drop this target entirely.

AlanCoding replied:
Due to path issues, I still strongly want some form of:

```
make test_modules MODULE_TEST_DIRS=awx_modules/test/awx/test_organization.py
```

to be available for development.

Adding the install and flake8 step to this target doesn't work with that goal to do quick single-test runs.

The naming of the combined target is complicated by the reference from https://github.com/ansible/zuul-jobs/pull/35

elyezer replied:
So this is needed for testing the modules isolated, right?

elyezer replied:
Talked offline with @AlanCoding and he explained that the intetion to have 3 make targets are:

* To be able to run only the prepare environment
* To be able to run only the tests for the modules
* To do both, prepare the environment and run the test with a single make target

Initially I thought having those 3 wouldn't be necessary but we actually need the three targets.

elyezer commented on Makefile:
```diff
@@ -377,8 +378,26 @@ test:
 	fi; \
 	PYTHONDONTWRITEBYTECODE=1 py.test -p no:cacheprovider -n auto $(TEST_DIRS)
 	cd awxkit && $(VENV_BASE)/awx/bin/tox -re py2,py3
+	cd /awx_devel
```
If this is a requirement for running the `test_modules` or `prepare_modules_venv` targets I would suggest moving this line inside that target.

elyezer replied:
Please never mind about this, more information here https://github.com/ansible/awx/pull/4701/files#r324263114

elyezer commented on awx_modules/galaxy.yml:
```diff
@@ -0,0 +1,16 @@
+authors:
+- Wayne Witzel III (@wwitzel3)
+- Chris Meyers <cmeyers@redhat.com>
```
Maybe add you as an author?

elyezer commented on setup.cfg:
```diff
@@ -18,4 +18,4 @@ exclude=.tox,venv,awx/lib/site-packages,awx/plugins/inventory/ec2.py,awx/plugins
 max-line-length=160
 ignore=E201,E203,E221,E225,E231,E241,E251,E261,E265,E303,W291,W391,W293,E731,W504
-exclude=.tox,venv,awx/lib/site-packages,awx/plugins/inventory,awx/ui,awx/api/urls.py,awx/main/migrations,awx/main/tests/data,node_modules/,awx/projects/,tools/docker,awx/settings/local_*.py,installer/openshift/settings.py,build/,installer/,awxkit/test
+exclude=.tox,venv,awx/lib/site-packages,awx/plugins/inventory,awx/ui,awx/api/urls.py,awx/main/migrations,awx/main/tests/data,node_modules/,awx/projects/,tools/docker,awx/settings/local_*.py,installer/openshift/settings.py,build/,installer/,awxkit/test,awx_modules/
```
I would suggest linting the modules code at least and just skip the tests. Actually I personally am a big fan of linting everything (including tests).

AlanCoding replied:
I have a `setup.cfg` file for the modules, and it is passing flake8.

The problem here is that I can't use different `setup.cfg` files for different folders. I still need to figure this out.

Ideally, I would just add the flake8 command for the modules to the `make flake8` command. But this command has an extremely non-obvious configuration and I frankly don't think this can be done cleanly.

That leaves with the option of maybe adding it to `make prepare_test_modules` or `make test_modules`. The overall configuration of these targets still needs a finalized design.

elyezer replied:
Got it, I was just making sure we were not caring about those files.

elyezer reviewed (commented):
Left some small comments, other than that this looks good to me.

AlanCoding commented:
From @ryanpetrello we sort of have a new proposal here: Rename everything that used to say something regarding "tower modules" to "awx collection".

Saying "modules" already has a major shortcoming due to the inclusion of the inventory plugin (which is not a module).

The makefile targets would become things like `make test_collection`. Much of the readme will need rework.

AlanCoding commented:
### Rebase notice

This is no ordinary rebase.

Before, I lifted content directly from https://github.com/AlanCoding/ansible-tower-content-export via a `git merge` (hand waving prior steps). If you dived into the commits, you would notice that this was a branch that did not share a common origin with the rest of AWX. I thought that might be a blocker for git management of this repo. Thinking of @matburt @ryanpetrello here.

First, I pulled in 2 new PRs from Ansible core that landed since I opened this PR (see in the content-export repo).

Then I squashed my own commits into 2. This is because I like to keep commits that move files separate from other commits (f193d76c322bcfb3cd283f47522e4d70a8d41704 is a doozy, it is not folder A to folder B).

Next, I rebased the `devel` branch from my content-export repo on top of the AWX `devel` branch. You can now see that this commit 5e68a75ccb183741592b1d15580697ffd1ec089c has a parent commit. This changed the commits to dual-authorship, which is honest.

With that done, I cherry picked the 2 commits of this PR. No conflicts were seen there. I did a test-rebase of the prior "merged" branch, and confirmed that only the only content difference was the new content from Ansible core. Then I fully replaced the prior branch with that branch.

That should, I think, alleviate any risks or concerns about merging this from the AWX git management standpoint.

AlanCoding commented:
One it lands, https://github.com/ansible/ansible/pull/62157 will add another thing to cherry pick. I do not see any need to insert it as an old commit, however, since it is my own commit.

AlanCoding committed:
Rename to collection, add license, galaxy build

AlanCoding committed:
rename awx modules folder to collection

AlanCoding committed:
update references to collection folder

AlanCoding committed:
Use namespaced doc fragment, cleanup

doc fragment will now be at awx.awx.auth
changed from just tower, which source from core

remove Makefile things no longer needed

AlanCoding commented on Makefile:
```diff
@@ -373,8 +374,37 @@ test:
 	fi; \
 	PYTHONDONTWRITEBYTECODE=1 py.test -p no:cacheprovider -n auto $(TEST_DIRS)
 	cd awxkit && $(VENV_BASE)/awx/bin/tox -re py2,py3
+	make test_collection_all
```
we will also remove this line once the test runs and confirms the new task is working.

AlanCoding commented on Makefile:
```diff
@@ -373,8 +374,37 @@ test:
 	awx-manage check_migrations --dry-run --check  -n 'vNNN_missing_migration_file'
 
+prepare_collection_venv:
+	cd /awx_devel
```
also remove

(I think it's implied you already must be there to run the command)

AlanCoding commented on Makefile:
```diff
@@ -373,8 +374,37 @@ test:
+test_collection:
+	@if [ "$(VENV_BASE)" ]; then \
+		. $(VENV_BASE)/awx/bin/activate; \
+	fi; \
```
@rooftopcellist I think you were doing some work that would allow us to ditch this repeated syntax. I'm all for that, and will be happy to make the related changes to this target as that shapes up.

shanemcd commented:
recheck

AlanCoding commented:
recheck

AlanCoding committed:
rename playbook vars to have collection_

AlanCoding commented:
Just a heads up, I have pushed a beta version to https://galaxy.ansible.com/awx/awx

This will be its ultimate home. After we've merged this, we will release an initial non-beta version. I feel like I would be happy to do 7.0.0 manually. This would be the point of parity with the old modules.

ryanpetrello reviewed (approved):
This is awesome, especially the Zuul CI.  Great work on this 😄
